File: Question/54_Spiral_Matrix.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        if(matrix.empty() || matrix[0].empty()) return {};
        int m = matrix.size();
        int n = matrix[0].size();
        int rounds = (min(m, n) + 1) / 2;
        vector<int> ans;

        int nx = n;
        int ny = m;
        for(int i = 0; i < rounds; ++i, nx -= 2, ny -=2) {
            for(int col = i; col < i + nx; ++col) {
                ans.push_back(matrix[i][col]);
            }
            for(int row = i + 1; row < i + ny; ++row) {
                ans.push_back(matrix[row][i + nx - 1]);
            }
            if(nx == 1 || ny == 1) break;

            for(int col = i + nx - 2; col >= i; --col) {
                ans.push_back(matrix[i + ny - 1][col]);
            }
            for(int row = i + ny - 2; row >= i + 1; --row) {
                ans.push_back(matrix[row][i]);
            }
        }
        return ans;
    }
};

File: Question_Java/54_Spiral_Matrix.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        List<Integer> ans = new ArrayList<>();
        int ys = 0;
        int ye = matrix.length - 1;
        int xs = 0;
        int xe = matrix[0].length - 1;
        while (ye >= ys && xe >= xs) {
            for (int i = xs; i <= xe; ++i) {
                ans.add(matrix[ys][i]);
            }
            ++ys;
            for (int i = ys; xs <= xe && i <= ye; ++i) {
                ans.add(matrix[i][xe]);
            }
            --xe;
            for (int i = xe; ys <= ye && i >= xs; --i) {
                ans.add(matrix[ye][i]);
            }
            --ye;
            for (int i = ye; xs <= xe && i >= ys; --i) {
                ans.add(matrix[i][xs]);
            }
            ++xs;
        }
        return ans;
    }
}